### PR TITLE
File path is in IO exception messages

### LINKED ISSUE #23710: Ensure file path is in IO exception messages
Historically IO exceptions have often not contained the path often due to CAS. That is not a concern any more and it is certainly more helpful to get "Access to the path '{0}' is denied." rather than "Access to the path is denied."

We fixed this in Corelib https://github.com/dotnet/coreclr/commit/6ba74dc2a7194f8d6c86c3aeab572a074ef645c8  .. we should do a similar pass over CoreFX. 

Look for strings like `Could not find a part of the path` `The specified file name or path is too long, or a component of the specified path is too long.` `Access to the path is denied` and insert the path where possible. Similarly `Illegal characters in path.`

### PR DESCRIPTION
**Pull request by satano**
Fixes #23710.

### CODE REVIEW

File: src/Common/src/System/IO/Win32Marshal.cs [hunk 1]
```diff
@@ -71,7 +71,7 @@ internal static Exception GetExceptionForWin32Error(int errorCode, string path)
                     return new IOException(SR.Format(SR.IO_AlreadyExists_Name, path), MakeHRFromErrorCode(errorCode));
 
                 case Interop.Errors.ERROR_FILENAME_EXCED_RANGE:
-                    return new PathTooLongException(SR.IO_PathTooLong);
+                    return new PathTooLongException(SR.Format(SR.IO_PathTooLong, path));
```
Comment (jkotas) [inline src/Common/src/System/IO/Win32Marshal.cs]: This should be done same way as it is done in CoreCLR
https://github.com/dotnet/coreclr/blob/6ba74dc2a7194f8d6c86c3aeab572a074ef645c8/src/mscorlib/shared/Interop/Unix/Interop.IOErrors.cs#L140

File: src/Common/src/System/IO/PathInternal.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ internal static void CheckInvalidPathChars(string path)
                 throw new ArgumentNullException(nameof(path));
 
             if (HasIllegalCharacters(path))
-                throw new ArgumentException(SR.Argument_InvalidPathChars, nameof(path));
+                throw new ArgumentException(SR.Format(SR.Argument_InvalidPathChars, path), nameof(path));
```
Comment (jkotas) [inline src/Common/src/System/IO/PathInternal.cs]: `Argument_InvalidPathChars` message does not have argument reference, so the path won't show up in the message.

Comment (satano) [reply src/Common/src/System/IO/PathInternal.cs]: Fixed.

**Comment by satano** [comment]
They are - argument was added in the first commit (for example https://github.com/dotnet/corefx/pull/25195/files#diff-52bde7ba793da54c362d3f5990017d2cR313). The only place I missed was in `Common` project you mentioned.

**Comment by jkotas** [comment]
Ok - I have missed it. It would be nice to do the matching change for the Argument_InvalidPathChars in https://github.com/dotnet/coreclr/ as well.

**Comment by satano** [comment]
@dotnet-bot test Tizen armel Debug Build

File: src/Common/tests/Resources/Strings.resx [hunk 3]
```diff
@@ -134,4 +153,7 @@
   <data name="IO_SharingViolation_File" xml:space="preserve">
     <value>IO_SharingViolation_File {0}</value>
   </data>
+  <data name="IO_PathTooLong_Path" xml:space="preserve">
```
Comment (danmosemsft) [inline src/Common/tests/Resources/Strings.resx]: I can't see coreclr (on phone) but hopefully the same strings are used in these changes: I rewrote some when I did the coreclr change, to be more helpful

File: src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs [hunk 4]
```diff
@@ -77,7 +77,10 @@ internal static void WinIOError(int errorCode, String str)
                         throw new UnauthorizedAccessException(string.Format(SR.UnauthorizedAccess_IODenied_Path, str));
 
                 case Interop.Errors.ERROR_FILENAME_EXCED_RANGE:
-                    throw new PathTooLongException(SR.IO_PathTooLong);
+                    if (string.IsNullOrEmpty(str))
```
Comment (danmosemsft) [inline src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs]: Is it actually possible for str to be null or empty here?

Comment (danmosemsft) [reply src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs]: I can't think why path too long would be appropriate if the path was empty or null?

File: src/Common/tests/Resources/Strings.resx (hunk 3 above)
Comment (satano) [reply src/Common/tests/Resources/Strings.resx]: Yes. Actually, I took the string from coreclr.

File: src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs (hunk 4 above)
Comment (satano) [reply src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs]: > Is it actually possible for str to be null or empty here?

Well, I do not know this. But for all other cases there (path not found, access denied...) there is a check. Though the checks are only for empy string (`str.Length == 0`). The second thing is, that it is this way also in coreclr, as @jkotas pointed in his review. Though it is in `System/IO/Win32Marshal.cs`.

> I can't think why path too long would be appropriate if the path was empty or null?

So I am not sure what to do here. If to leave it as I did it, or change it to test only for empty string like the other cases, or revert it back. Personally, I would remove the check for empty stirng and left only new message with path. And I would make it the same way in `Win32Marshal.cs` file.

Comment (jkotas) [reply src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs]: > Is it actually possible for str to be null or empty here?

It is possible when the OS API returns the path too long error code, but the caller of this method does not have the path.

File: src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs [hunk 5]
```diff
@@ -77,7 +77,10 @@ internal static void WinIOError(int errorCode, String str)
...
+                        throw new PathTooLongException(SR.IO_PathTooLong);
+                    else
+                        throw new PathTooLongException(SR.Format(SR.IO_PathTooLong_Path, str));
```
Comment (stephentoub) [inline src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs]: Nit: it'd be nice to use braces here on both the if/else blocks, or else condense this into a single throw, e.g.
```C#
throw string.IsNullOrEmpty(str) ?
    new PathTooLongException(SR.IO_PathTooLong) :
    new PathTooLongException(SR.Format(SR.IO_PathTooLong_Path, str));
```

Comment (stephentoub) [reply src/System.IO.Ports/src/System/IO/Ports/InternalResources.cs]: (though I see now that it's just copying the style used elsewhere in the file, so ok)

**Pull request merged**